spring-cloud-aws-sqs/src/main/java/io/awspring/cloud/sqs/listener/SqsMessageProducer.java

```diff
@@ -0,0 +1,160 @@
+import java.util.function.Function;
+import java.util.stream.Collectors;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.context.SmartLifecycle;
+import org.springframework.messaging.Message;
+import org.springframework.messaging.support.GenericMessage;
+import org.springframework.util.MimeType;
+import software.amazon.awssdk.services.sqs.SqsAsyncClient;
+import software.amazon.awssdk.services.sqs.model.MessageSystemAttributeName;
+import software.amazon.awssdk.services.sqs.model.ReceiveMessageResponse;
+
+/**
+ * @author Tomaz Fernandes
+ * @since 3.0
+ */
+public class SqsMessageProducer implements AsyncMessageProducer<String>, SmartLifecycle {
+
+	private static final Logger logger = LoggerFactory.getLogger(SqsMessageProducer.class);
+
+	private final String logicalEndpointName;
+
+	private final QueueAttributes queueAttributes;
+
+	private final SqsAsyncClient sqsAsyncClient;
+
+	private final String queueUrl;
+
+	private volatile boolean running;
+
+	public SqsMessageProducer(String logicalEndpointName, QueueAttributes queueAttributes, SqsAsyncClient sqsClient) {
+		this.logicalEndpointName = logicalEndpointName;
+		this.queueUrl = queueAttributes.getDestinationUrl();
+		this.queueAttributes = queueAttributes;
+		this.sqsAsyncClient = sqsClient;
+	}
+
+	@Override
+	public CompletableFuture<Collection<Message<String>>> produce(int numberOfMessages, Duration timeout) {
+		logger.trace("Polling for messages at {}", this.queueUrl);
+		return sqsAsyncClient
+				.receiveMessage(req -> req.queueUrl(this.queueUrl).maxNumberOfMessages(numberOfMessages)
+						.waitTimeSeconds((int) timeout.getSeconds()))
+				.thenApply(ReceiveMessageResponse::messages).thenApply(this::getMessagesForExecution)
+				.exceptionally(handleException());
+	}
+
+	protected Function<Throwable, Collection<Message<String>>> handleException() {
+		return t -> {
+			logger.error("Error retrieving messages from SQS", t);
+			return Collections.emptyList();
+		};
+	}
+
+	private Collection<Message<String>> getMessagesForExecution(
+			List<software.amazon.awssdk.services.sqs.model.Message> messages) {
+		logger.trace("Poll returned {} messages", messages.size());
+		return messages.stream().map(this::getMessageForExecution).collect(Collectors.toList());
+	}
+
+	protected org.springframework.messaging.Message<String> getMessageForExecution(
+			final software.amazon.awssdk.services.sqs.model.Message message) {
+		HashMap<String, Object> additionalHeaders = new HashMap<>();
+		additionalHeaders.put(SqsMessageHeaders.SQS_LOGICAL_RESOURCE_ID, this.logicalEndpointName);
+		additionalHeaders.put(SqsMessageHeaders.RECEIVED_AT, Instant.now());
+		additionalHeaders.put(SqsMessageHeaders.QUEUE_VISIBILITY, this.queueAttributes.getVisibilityTimeout());
+		additionalHeaders.put(SqsMessageHeaders.VISIBILITY,
+				new QueueMessageVisibility(this.sqsAsyncClient, this.queueUrl, message.receiptHandle()));
+		return createMessage(message, additionalHeaders);
```

Comment:
I would rely on `Collections.unmodifiableMap()`
```suggestion
		return createMessage(message, Collections.unmodifiableMap(additionalHeaders));
```